### ISSUE TITLE
The performance is worse than box2d.js

### ISSUE THREAD

**Issue by finscn** [liabru/matter-js#608]
I found the performance of matter.js is worse than box2d.js  (only 40%)

There is a test I created:

http://fatidol.com/phy-benchmark/

There are many different box2d.js projects , I chose https://github.com/flyover/box2d.js ,
Because it includes Google's LiquidFun , I like this feature.

**Comment by bchevalier** [comment]
Hey @finscn, could you try out one of my commits? I heavily optimized the engine
https://github.com/liabru/matter-js/pull/566/commits/edc7c03dab0d540171feefde2be10efd2545544f
Performance may be roughly 10x better.

**Comment by finscn** [comment]
@bchevalier , hi , could you give me a matter.min.js file directly ?  I don't know how to build

**Comment by bchevalier** [comment]
Here you go: [minified](https://github.com/bchevalier/matter-js/blob/optimized-build-test/build/matter.min.js) and [unminified](https://github.com/bchevalier/matter-js/blob/optimized-build-test/build/matter.js)

**Comment by bchevalier** [comment]
I noticed that setting both iterations to ~12 was providing good result. Would you mind updating your online benchmark with the changes? I am curious to see what it gives.

**Comment by finscn** [comment]
I tried.  Even I set iterations to 30 , 
![image](https://user-images.githubusercontent.com/288367/39398964-4283789e-4b48-11e8-9e7b-92cb24621825.png)

But there are still many rect-boxes are overlap .

BTW , 
Your version is very fast. 
I set the count of rect-box to 800 ,  it's still keeping  60 fps on my macbook,  
the box2d.ts  is down to 40 fps.

**Comment by bchevalier** [comment]
Hey @liabru, do you have any idea what may be the reason for that difference in behavior between matter and box2d? I am willing to have a look and try to fix the issue.

**Comment by liabru** [comment]
Indeed @bchevalier has done some excellent performance work that I've yet to fully integrate (I'll be working on it in May!).

There are likely a few reasons for this, one being the runner does not use a fixed delta by default and the low FPS causing a death spiral.

Try `runner.isFixed = true` and `runner.delta = 5` which should help a bit.

The runner has a lot of room for improvement here which is something I need to revisit.

Other improvements for this would be adaptive iterations (minimum error threshold) which I have a mostly-working implementation (it can cause some wild behaviour).

Also is improving the resolvers initial 'guess' when a pair first collides. In other words, predictive collision warming (warming is already implemented, but it is not primed with any initial guess). The guess would be based on velocity and mass.

@bchevalier the latter is something I can fill you in on some more if you'd like to tackle it!

**Comment by finscn** [comment]
@liabru , thanks for your suggestion.
I've tried runner.isFixed = true and runner.delta = 5 , it's better , but still worse than box2d.

I'll be waiting for the improvement. Thank you

**Comment by bchevalier** [comment]
@liabru yes please fill me in on that if you don't mind. I cannot promise anything in terms of timeline though.

**Comment by finscn** [comment]
Hi  @bchevalier & @liabru  , is there any news about new version ?

**Comment by yungzhu** [comment]
Looking forward to performance improvements
Is there any plan or time for this regard?

**Comment by jcyuan** [comment]
I found that you added separating axis support for Body after comparing between your version and master's.
@bchevalier @liabru

**Comment by finscn** [comment]
Hi @@flyover , I think @jcyuan 's question is a good question.
@bchevalier 's forked matter-js is faster than box2d.ts . (original matter-js is not)
I wish box2d.ts could as fast as @bchevalier 's forked matter-js .

**Comment by jcyuan** [comment]
warmly ping @bchevalier @liabru 
I think the web physics engine still activating is getting more less especially which suitable for mobile client. hope this project can be better and better. thanks for your efforts.

**Comment by bchevalier** [comment]
@jcyuan, sorry I did not notice your question back in July.
To make it faster than box2d, mainly 2 things:
- JavaScript specific optimization (make sure the JS code does not get de-optimized/can be optimized by the JS engine compiler/optimizer).
- Decreasing the complexity of some methods so that for instance their complexity goes from O(n^2) to O(n)

These optimizations do not require any knowledge specific to physics engines.

I would imagine that it's possibly to optimize box2d as well (because the great majority of projects can be optimized).

**Comment by jcyuan** [comment]
@bchevalier thanks for the info, could you please provide some example? I dont quite understand what kind of optimization can make a such fps difference.

**Comment by mrzmyr** [comment]
@liabru any updates on this?

**Comment by mrzmyr** [comment]
@liabru any idea when to merge it ?

**Comment by mattbucci** [comment]
@bchevalier and @liabru any idea when these performance updates will be merged? Anything we can do to help?

**Comment by jcyuan** [comment]
I'm using @flyover 's [box2d.ts](https://github.com/flyover/box2d.ts).
here is a [benchmark](http://fatidol.com/phy-benchmark/) made by @finscn

**Comment by mapacarta** [comment]
@bchevalier I tested your commit, it seems much faster. Thanks for your efforts.

In my game, I cancel some of the collisions with pair.isActive = false;  
It seems you removed this, is there any other way to do this?

**Comment by megasyl** [comment]
@bchevalier Thank you very much ! Rendering is 10 to 15 times faster with your optimizations ! Very nice work !

**Comment by Adamsmashem** [comment]
I found that doing ```Matter.Engine.clear(engine) ``` every frame helps a lot with only a slight accuracy drop 

here is an example 

press space to not do ```Matter.Engine.clear(engine) ```

https://www.openprocessing.org/sketch/932385
